src/main/scala/firrtl/transforms/ManipulateNames.scala

```diff
@@ -0,0 +1,249 @@
+    implicit val mType: ModuleType = modType
+
+    // Store local renames of refs to instances to their renamed modules. This is needed when renaming port connections
+    // on subfields where only the local instance name is available.
+    implicit val iToM: mutable.Map[ComponentName, ModuleName] = mutable.Map.empty
+
+    val mx = m
+      .map(onPort)
+      .map(onStatement)
+      .map(onName(_: String)(renames, moduleNamespace, Some(circuit)))
+
+    // Must happen after renaming the name and ports of the module itself
+    mType += (mx.name -> onType(Utils.module_type(mx)))
+    mx
+  }
+
+  /** Fix any Verilog keyword collisions in a [[firrtl.ir Circuit]]
+    * @param c a [[firrtl.ir Circuit]] with possible name collisions
+    * @param renames a [[RenameMap]] to update. If you don't want to propagate renames, this can be ignored.
+    * @return a [[firrtl.ir Circuit]] without keyword conflicts
+    */
+  def run(c: ir.Circuit, renames: RenameMap = RenameMap()): ir.Circuit = {
+    implicit val circuitNamespace: Namespace = Namespace(c)
+    implicit val scope: Option[CircuitName] = Some(CircuitName(c.main))
+    val modType: ModuleType = new ModuleType()
+
+    // Rename all modules from leafs to root in one pass while updating a shared rename map. Going from leafs to roots
+    // ensures that the rename map is safe for parents to blindly consult.
+    val modulesx: Map[ModuleName, Seq[ir.DefModule]] = new InstanceGraph(c).moduleOrder.reverse
+      .map(onModule(renames, scope.get, modType))
+      .groupBy(m => ModuleName(m.name, scope.get))
+
+    // Reorder the renamed modules into the original circuit order.
+    val modulesxx: Seq[ir.DefModule] = c.modules.flatMap{ orig =>
+      val named = ModuleName(orig.name, scope.get)
+      modulesx(renames.get(named).orElse(Some(Seq(named))).get.head)
+    }
+
+    // Rename the circuit if the top module was renamed
+    val mainx = renames.get(ModuleName(c.main, CircuitName(c.main))) match {
+      case Some(Seq(ModuleName(m, _))) =>
+        renames.rename(CircuitName(c.main), CircuitName(m))
+        m
+      case x@ Some(_) => throw new PassException(
+        s"Verilog renaming shouldn't result in multiple renames, but found '${c.main} -> $x'")
+      case None =>
+        c.main
+    }
+
+    // Apply all updates
+    c.copy(modules = modulesxx, main = mainx)
+  }
+
+  /** Fix any Verilog keyword name collisions in a [[CircuitState]] while propagating renames
+    * @param state the [[CircuitState]] with possible name collisions
+    * @return a [[CircuitState]] without name collisions
+    */
+  def execute(state: CircuitState): CircuitState = {
+    val renames = RenameMap()
+    renames.setCircuit(state.circuit.main)
+    state.copy(circuit = run(state.circuit, renames), renames = Some(renames))
+  }
+
+}
+
+/** Parent of transforms that do change the letter case of names in a FIRRTL circuit */
+abstract class LetterCaseTransform extends ManipulateNames {
```

Comment:
Is it worth thinking about a new package for new "feature" transforms? Moving all the core lowering transforms into a core lowering package will be something of a pain point, but it might be good to start something like `firrtl.features` or something.

Reply:
I think we can do the improvement later, the only necessary change for now would be reflecting the `GetNamespace` invalidation.

Reply:
Since this is 1.3.x, it's probably better to start the new package now as opposed to this getting into 1.3.x and then us needing to go through a deprecation cycle to move `LetterCaseTransform`. I'm good with `firrtl.features`. Any idea of a convention that compilers use for the separation of core vs. X?

Reply:
Updated in the rebase to add `firrtl.features`.